cpp/benchmarks/common/generate_input.hpp

```diff
@@ -389,16 +390,120 @@ class data_profile {
     struct_dist_desc.max_depth = max_depth;
   }
 
-  void set_struct_types(std::vector<cudf::type_id> const& types)
+  void set_struct_types(cudf::host_span<cudf::type_id const> types)
   {
     CUDF_EXPECTS(
       std::none_of(
-        types.cbegin(), types.cend(), [](auto& type) { return type == cudf::type_id::STRUCT; }),
+        types.begin(), types.end(), [](auto& type) { return type == cudf::type_id::STRUCT; }),
       "Cannot include STRUCT as its own subtype");
-    struct_dist_desc.leaf_types = types;
+    struct_dist_desc.leaf_types.assign(types.begin(), types.end());
   }
 };
 
+/**
+ * @brief Builder to construct data profiles for the random data generator.
+ *
+ * Setters can be chained to set multiple properties in a single expression.
+ * For example, `data_profile` initialization
+ * @code{.pseudo}
+ * data_profile profile;
+ * profile.set_null_frequency(0.0);
+ * profile.set_cardinality(0);
+ * profile.set_distribution_params(cudf::type_id::INT32, distribution_id::UNIFORM, 0, 100);
+ * @endcode
+ * becomes
+ * @code{.pseudo}
+ * data_profile const profile =
+ *   data_profile_builder().cardinality(0).null_frequency(0.0).distribution(
+ *     cudf::type_id::INT32, distribution_id::UNIFORM, 0, 100);
+ * @endcode
+ * The builder makes it easier to have immutable `data_profile` objects even with the complex
+ * initialization. The `profile` object in the example above is initialized from
+ * `data_profile_builder` using an implicit conversion operator.
+ *
+ * The builder API also includes a few additional convinience setters:
+ * Overload of `distribution` that only takes the distribution type (not the range).
+ * `no_validity`, which is a simpler equivalent of `null_frequency(std::nullopr)`.
+ */
+class data_profile_builder {
+  data_profile profile;
+
+ public:
+  template <typename T, typename Type_enum>
+  data_profile_builder& distribution(Type_enum type_or_group, distribution_id dist)
+  {
+    auto const range = default_range<T>();
+    profile.set_distribution_params(type_or_group, dist, range.first, range.second);
+    return *this;
+  }
+
+  template <class... T>
+  data_profile_builder& distribution(T&&... t)
+  {
+    profile.set_distribution_params(std::forward<T>(t)...);
+    return *this;
+  }
+
+  data_profile_builder& bool_probability(double p)
```

Comment:
Could we have some doxygen here that explains each of these member functions a bit?

Reply:
Added docs and realized that naming was... bad (trying to explain something really brings clarity :)). Renamed a few functions, should be more consistent/correct/clear now.